docs/SAFETY-MACROS.md

```diff
@@ -0,0 +1,351 @@
+
+
+### POSIX_CHECKED_MEMSET(destination, value, len)
+
+Performs a safe memset
+
+
+### POSIX_GUARD(result)
+
+Ensures `(result) >= S2N_SUCCESS`, otherwise the function will return `S2N_FAILURE`
+
+
+### POSIX_GUARD_OSSL(result, error)
+
+Ensures `result == _OSSL_SUCCESS`, otherwise the function will `POSIX_BAIL` with `error`
+
+
+### POSIX_GUARD_RESULT(result)
+
+Ensures `s2n_result_is_ok(result)`, otherwise the function will return `S2N_FAILURE`
+
+
+### POSIX_GUARD_POSIX(result)
+
+Ensures `(result) >= S2N_SUCCESS`, otherwise the function will return `S2N_FAILURE`
+
+
+### POSIX_GUARD_PTR(result)
+
+Ensures `(result) != NULL`, otherwise the function will return `S2N_FAILURE`
+
+
+## Macros for functions that return a pointer
+
+
+### PTR_BAIL(error)
+
+Sets the global `s2n_errno` to `error` and returns with an `NULL`
+
+
+### PTR_ENSURE(condition, error)
+
+Ensures the `condition` is `true`, otherwise the function will `PTR_BAIL` with `error`
+
+
+### PTR_ENSURE_OK(result, error)
+
+Ensures `(result) != NULL`, otherwise the function will `PTR_BAIL` with `error`
+
+        This can be useful for overriding the global `s2n_errno`
+
+
+### PTR_ENSURE_GTE(a, b)
+
+Ensures `a` is greater than or equal to `b`, otherwise the function will `PTR_BAIL` with a `S2N_ERR_SAFETY` error
+
+
+### PTR_ENSURE_LTE(a, b)
+
+Ensures `a` is less than or equal to `b`, otherwise the function will `PTR_BAIL` with a `S2N_ERR_SAFETY` error
+
+
+### PTR_ENSURE_GT(a, b)
+
+Ensures `a` is greater than `b`, otherwise the function will `PTR_BAIL` with a `S2N_ERR_SAFETY` error
+
+
+### PTR_ENSURE_LT(a, b)
+
+Ensures `a` is less than `b`, otherwise the function will `PTR_BAIL` with a `S2N_ERR_SAFETY` error
+
+
+### PTR_ENSURE_EQ(a, b)
+
+Ensures `a` is equal to `b`, otherwise the function will `PTR_BAIL` with a `S2N_ERR_SAFETY` error
+
+
+### PTR_ENSURE_NE(a, b)
+
+Ensures `a` is not equal to `b`, otherwise the function will `PTR_BAIL` with a `S2N_ERR_SAFETY` error
+
+
+### PTR_ENSURE_INCLUSIVE_RANGE(min, n, max)
+
+Ensures `min <= n <= max`, otherwise the function will `PTR_BAIL` with `S2N_ERR_SAFETY`
+
+
+### PTR_ENSURE_EXCLUSIVE_RANGE(min, n, max)
+
+Ensures `min < n < max`, otherwise the function will `PTR_BAIL` with `S2N_ERR_SAFETY`
+
+
+### PTR_ENSURE_REF(x)
+
+Ensures `x` is a readable reference, otherwise the function will `PTR_BAIL` with `S2N_ERR_NULL`
+
+
+### PTR_ENSURE_MUT(x)
+
+Ensures `x` is a mutable reference, otherwise the function will `PTR_BAIL` with `S2N_ERR_NULL`
+
+
+### PTR_PRECONDITION(condition)
+
+Ensures the `condition` is `true`, otherwise the function will `PTR_BAIL` with a `S2N_ERR_PRECONDITION_VIOLATION` error
+
+
+### PTR_POSTCONDITION(condition)
+
+Ensures the `condition` is `true`, otherwise the function will `PTR_BAIL` with a `S2N_ERR_POSTCONDITION_VIOLATION` error
+
+
+### PTR_CHECKED_MEMCPY(destination, source, len)
+
+Performs a safe memcpy
+
+
+### PTR_CHECKED_MEMSET(destination, value, len)
+
+Performs a safe memset
```

Comment:
Could we define "safe"?

Reply:
Will do.

Reply:
I think this a lot better: https://github.com/camshaft/s2n/blob/safety-macros/docs/SAFETY-MACROS.md#result_checked_memcpydestination-source-len

Also, after looking at the [`memcpy` documentation](http://www.cplusplus.com/reference/cstring/memcpy/) I think it's actually safer to use [`memmove`](http://www.cplusplus.com/reference/cstring/memmove/) instead, otherwise callers will need to ensure that their pointers don't overlap. If they do, memcpy is potential UB!